File: static/global/items.js
/* Copyright (C) 2021-2023 tiksan

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>. */

var items = null;

function _writeItems() {
    $.each(items, function (itemID, itemName) {
        $(".item-selector").append(
            $("<option>", {
                value: itemID,
                text: itemName,
            })
        );
    });
}

let itemsRequest = (obj) => {
    var localItems = JSON.parse(localStorage.getItem("items"));

    if (localItems && Math.floor(Date.now() / 1000) - localItems.timestamp < 3600) {
        return new Promise((resolve, reject) => {
            items = localItems.items;
            _writeItems();
            resolve();
        });
    } else {
        return new Promise((resolve, reject) => {
            let xhttp = new XMLHttpRequest();

            xhttp.onload = function () {
                let response = xhttp.response;

                if ("code" in response) {
                    generateToast("Torn Items Not Located", response["message"]);
                    reject();
                    return;
                }

                items = response.items;
                _writeItems();
                localStorage.setItem(
                    "items",
                    JSON.stringify({
                        timestamp: Math.floor(Date.now() / 1000),
                        items: items,
                    })
                );
                resolve();
            };

            xhttp.responseType = "json";
            xhttp.open("GET", "/api/items");
            xhttp.setRequestHeader("Content-Type", "application/json");
            xhttp.send();
        });
    }
};

File: celery/tornium_celery/tasks/items.py
# Copyright (C) 2021-2023 tiksan
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import datetime

from tornium_commons import rds
from tornium_commons.models import Item, User

import celery
from celery.utils.log import get_task_logger

from .api import tornget

logger = get_task_logger("celery_app")


@celery.shared_task(
    name="tasks.items.update_items",
    routing_key="default.items.update_items",
    queue="default",
    time_limit=15,
)
def update_items(items_data):
    Item.update_items(torn_get=tornget, key=User.random_key().api_key)

    rds().set(
        "tornium:items:last-update",
        int(datetime.datetime.utcnow().timestamp()),
        ex=5400,
    )  # 1.5 hours
